AirCasting/CoreData/MeasurementStreamStorage.swift

```diff
@@ -0,0 +1,121 @@
+// Created by Lunar on 09/05/2021.
+//
+
+import Foundation
+import CoreLocation
+import CoreData
+
+protocol MeasurementStreamStorage {
+    func addMeasurement(_ measurement: Measurement, toStreamWithID id: MeasurementStreamLocalID) throws
+    func createMeasurementStream(_ stream: MeasurementStream, for sessionUUID: SessionUUID) throws -> MeasurementStreamLocalID
+    func createSession(_ session: Session) throws
+    func createSessionAndMeasurementStream(_ session: Session, _ stream: MeasurementStream) throws -> MeasurementStreamLocalID
+    func updateSessionStatus(_ sessionStatus: SessionStatus, for sessionUUID: SessionUUID) throws
+}
+
+extension MeasurementStreamStorage {
+    func addMeasurementValue(_ value: Double, at location: CLLocationCoordinate2D? = nil, toStreamWithID id: MeasurementStreamLocalID) throws {
+        try addMeasurement(Measurement(time: Date(), value: value, location: location), toStreamWithID: id)
+    }
+}
+
+final class CoreDataMeasurementStreamStorage: MeasurementStreamStorage {
+    enum Error: Swift.Error {
+        case missingMeasurementStream
+    }
+    private let persistenceController: PersistenceController
+    private lazy var updateSessionParamsService = UpdateSessionParamsService()
+
+    init(persistenceController: PersistenceController) {
+        self.persistenceController = persistenceController
+    }
+
+    func addMeasurement(_ measurement: Measurement, toStreamWithID id: MeasurementStreamLocalID) throws {
+        let context = persistenceController.editContext()
```

Comment:
This is very important to have it created each time so the invalid written changes to the context failed to save it will not leave the viewContext in a invalid state. also nsmanageobjectcontext is connected to the current thread and creation of a context makes it safe threaded.